lib/elixir/src/elixir_errors.erl

```diff
@@ -22,18 +24,234 @@ print_warning(Location, File, Message) ->
+        [
+         Spacing, prefix(warning), file_format(Position, File),
+         Spacing, Message, 
+         Spacing, length(Stack),
+         Stack
+        ]
+      )
+  end,
+  io:put_chars(standard_error, Output).
+
+extract_line({L, _}) -> L;
+extract_line(L) -> L.
+
+no_line_diagnostic(Position, File, Message, Severity) -> 
+  io_lib:format(
+    " ┌─ ~ts~ts\n"
+    " ~ts",
+    [
+     prefix(Severity), file_format(Position, File),
+     Message
+    ]
+   ).
+
+line_column_diagnostic(Position, File, Message, Severity) ->
+  {LineNumber, Column} = Position,
+  Line = get_file_line(File, LineNumber),
+  {TrimmedLine, TotalLeading} = trim_file_line(Line),
+  LineDigits = get_line_number_digits(LineNumber),
+  Spacing = n_spaces(LineDigits + 1),
+
+  io_lib:format(
+    " ~ts┌─ ~ts~ts\n"
+    " ~ts│\n"
+    " ~p │ ~ts\n"
+    " ~ts│ ~ts\n\n"
+    " ~ts~ts",
+    [
+     Spacing, prefix(Severity), file_format(Position, File),
+     Spacing,
+     LineNumber, TrimmedLine, 
+     Spacing, highlight_below_line(TrimmedLine, Column - TotalLeading, Severity),
+     Spacing, format_message(Message, LineDigits)
+    ]
+  ).
+
+line_only_diagnostic(LineNumber, File, Message, Severity) ->
+  Line = get_file_line(File, LineNumber),
+  {TrimmedLine, _} = trim_file_line(Line),
+  LineDigits = get_line_number_digits(LineNumber),
+  Spacing = n_spaces(LineDigits + 1),
+
+  io_lib:format(
+    " ~ts┌─ ~ts~ts\n"
+    " ~ts│\n"
+    " ~p │ ~ts\n"
+    " ~ts│ ~ts\n\n"
+    " ~ts~ts",
+    [ 
+     Spacing, prefix(Severity), file_format(LineNumber, File),
+     Spacing,
+     LineNumber, TrimmedLine,
+     Spacing, highlight_below_line(TrimmedLine, Severity),
+     Spacing, format_message(Message, LineDigits)
+    ]
+ ).
+
+highlight_below_line(Line, Column, Severity) ->
+  ErrorLength = match_line_error(Line, Column),
+  case Severity of
+    warning ->  highlight([n_spaces(Column), lists:duplicate(ErrorLength, "~")], warning);
+    error -> highlight([n_spaces(Column), lists:duplicate(ErrorLength, "^")], error)
+  end.
+
+highlight_below_line(Line, Severity) ->
+  Length = string:length(Line),
+  case Severity of 
+    warning -> highlight(lists:duplicate(Length, "~"), warning);
+    error -> highlight(lists:duplicate(Length, "^"), error)
+  end.
+
+get_line_number_digits(Number) -> do_get_line_number_digits(Number, 1).
+
+do_get_line_number_digits(Number, Acc) when Number < 10 -> Acc;
+do_get_line_number_digits(Number, Acc) -> 
+  do_get_line_number_digits(Number div 10, Acc + 1).
+
+match_line_error(Line, Column) ->
+  TermRegex = "[A-Za-z_\.\{\}\(\)&@]*",
+  {ok, Re} = re:compile(TermRegex),
+  Tail = string:slice(Line, Column),
+ {match, [MatchingTerm]} = re:run(Tail, Re, [{capture, all, binary}]),
+
+  case MatchingTerm of 
+    % If we cannot consume anything, return a single ^ at column position
+    <<>> -> 1;
+    _ -> string:length(MatchingTerm)
+  end.
+
+get_file_line(File, LineNumber) -> 
+  {ok, Data} = file:read_file(File),
```

Comment:
Here we are reading the whole file just to get the line, which is not ideal, I was using this code snippet before:

```erl
  {ok, IoDevice} = file:open(File, [read]),
  LineCollector = fun 
                    (I, nil) when I == LineNumber - 1 -> 
                      {ok, Line} = file:read_line(IoDevice),
                      string:trim(Line, trailing);
                    (_, nil) -> 
                      {ok, _} = file:read_line(IoDevice),
                      nil;
                    (_, Line) -> 
                      Line
                  end,
  Line = lists:foldl(LineCollector, nil, lists:seq(0, LineNumber)),
  file:close(IoDevice),
  Line.
```

But I couldn't get it working with unicode after a lot of adaptations, so ideally we should make it handle unicode and use it.